File: vol1/java-examples/src/main/java/com/heatonresearch/aifh/examples/learning/LearnPolynomial.java
/*
 * Artificial Intelligence for Humans
 * Volume 1: Fundamental Algorithms
 * Java Version
 * http://www.aifh.org
 * http://www.jeffheaton.com
 *
 * Code repository:
 * https://github.com/jeffheaton/aifh

 * Copyright 2013 by Jeff Heaton
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * For more information on Heaton Research copyrights, licenses
 * and trademarks visit:
 * http://www.heatonresearch.com/copyright
 */

package com.heatonresearch.aifh.examples.learning;

import com.heatonresearch.aifh.general.data.BasicData;
import com.heatonresearch.aifh.learning.TrainGreedyRandom;
import com.heatonresearch.aifh.learning.score.ScoreFunction;
import com.heatonresearch.aifh.learning.score.ScoreRegressionData;

import java.util.ArrayList;
import java.util.List;

/**
 * Learn a simple polynomial with the Greedy Random algorithm.
 */
public class LearnPolynomial extends SimpleLearn {

    public List<BasicData> generateTrainingData() {
        final List<BasicData> result = new ArrayList<BasicData>();

        for (double x = -50; x < 50; x++) {
            final double y = (2 * Math.pow(x, 2)) + (4 * x) + 6;
            final BasicData pair = new BasicData(1, 1);
            pair.getInput()[0] = x;
            pair.getIdeal()[0] = y;
            result.add(pair);
        }

        return result;
    }


    /**
     * Run the example.
     */
    public void process() {
        final List<BasicData> trainingData = generateTrainingData();
        final PolynomialFn poly = new PolynomialFn(3);
        final ScoreFunction score = new ScoreRegressionData(trainingData);
        final TrainGreedyRandom train = new TrainGreedyRandom(true, poly, score);
        performIterations(train, 1000000, 0.01, true);
        System.out.println(poly.toString());
    }

    /**
     * The main method.
     *
     * @param args Not used.
     */
    public static void main(final String[] args) {
        final LearnPolynomial prg = new LearnPolynomial();
        prg.process();
    }
}

File: vol1/csharp-examples/AIFH-Vol1/Examples/Learning/LearnPolynomial.cs
using System;
using System.Collections.Generic;
using AIFH_Vol1.Core.General.Data;
using AIFH_Vol1.Core.Learning;
using AIFH_Vol1.Core.Learning.Score;

namespace AIFH_Vol1.Examples.Learning
{
    /// <summary>
    /// Learn a simple polynomial with the Greedy Random algorithm.
    /// </summary>
    public class LearnPolynomial : SimpleLearn
    {
        /// <summary>
        /// The name of this example.
        /// </summary>
        public static string ExampleName = "Learn a simple polynomial with the Greedy Random algorithm";

        /// <summary>
        /// The chapter this example is from.
        /// </summary>
        public static int ExampleChapter = 7;

        public IList<BasicData> GenerateTrainingData()
        {
            IList<BasicData> result = new List<BasicData>();

            for (double x = -50; x < 50; x++)
            {
                double y = (2 * Math.Pow(x, 2)) + (4 * x) + 6;
                var pair = new BasicData(1, 1);
                pair.Input[0] = x;
                pair.Ideal[0] = y;
                result.Add(pair);
            }

            return result;
        }


        /// <summary>
        /// Run the example.
        /// </summary>
        public void Process()
        {
            IList<BasicData> trainingData = GenerateTrainingData();
            var poly = new PolynomialFn(3);
            IScoreFunction score = new ScoreRegressionData(trainingData);
            var train = new TrainGreedyRandom(true, poly, score);
            PerformIterations(train, 1000000, 0.01, true);
            Console.WriteLine(poly.ToString());
        }

        /// <summary>
        ///     The entry point for this example.  If you would like to make this example
        ///     stand alone, then add to its own project and rename to Main.
        /// </summary>
        /// <param name="args">Not used.</param>
        public static void ExampleMain(string[] args)
        {
            var prg = new LearnPolynomial();
            prg.Process();
        }
    }
}
